File: iosApp/iosApp/presentation/notes/NotesScreen.swift
import SwiftUI

struct NotesScreen: View {
    @Environment(\.colorScheme) private var colorScheme
    let mode: NotesScreenMode
    let summaryText: String
    let statusLabel: String
    let filters: [NotesSourceFilterItem]
    let memos: [MemoListItem]
    let onRecordClick: () -> Void
    let onSelectSourceFilter: (String) -> Void
    let onRenameMemo: (String) -> Void
    let onDeleteMemo: (String) -> Void

    var body: some View {
        ZStack(alignment: .bottom) {
            ScrollView(showsIndicators: false) {
                VStack(alignment: .leading, spacing: 14) {
                    NotesHeader(
                        titleColor: titleColor,
                        metaColor: metaColor,
                        summaryText: summaryText,
                        statusLabel: statusLabel
                    )
                    NotesSearchBar(textColor: metaColor, background: searchBackground, border: searchBorder)

                    if mode == .loading {
                        NotesLoadingFilterRow(
                            background: sourceChipBackground,
                            selectedBackground: titleColor
                        )

                        NotesLoadingStatusCard(
                            titleColor: titleColor,
                            subtitleColor: metaColor,
                            background: cardBackground,
                            border: cardBorder,
                            iconColor: metaColor,
                            accentColor: statusChipAccent,
                            skeletonColor: sourceChipBackground
                        )

                        VStack(spacing: 12) {
                            ForEach(0..<4, id: \.self) { _ in
                                LoadingMemoCard(
                                    background: cardBackground,
                                    border: cardBorder,
                                    lineColor: searchBorder,
                                    chipColor: sourceChipBackground
                                )
                            }
                        }
                    } else if mode == .empty {
                        NotesEmptyStateCard(
                            titleColor: titleColor,
                            subtitleColor: metaColor,
                            background: emptyStateBackground,
                            border: emptyStateBorder,
                            iconColor: emptyStateIconColor
                        )
                        .padding(.top, 96)
                    } else {
                        NotesFilterRow(
                            filters: filters,
                            onSelect: onSelectSourceFilter
                        )

                        NotesRecentLabel(textColor: metaColor)

                        VStack(spacing: 12) {
                            ForEach(memos) { memo in
                                MemoCard(
                                    memo: memo,
                                    background: cardBackground,
                                    border: cardBorder,
                                    titleColor: titleColor,
                                    metaColor: metaColor,
                                    sourceBackground: sourceChipBackground,
                                    sourceTextColor: sourceChipText,
                                    onRename: { onRenameMemo(memo.id) },
                                    onDelete: { onDeleteMemo(memo.id) }
                                )
                            }
                        }
                    }

                    Spacer()
                        .frame(height: 92)
                }
                .padding(.horizontal, VoraSpacing.pageHorizontal)
                .padding(.top, VoraSpacing.pageTop)
                .padding(.bottom, VoraSpacing.pageBottom)
            }

            RecordButton(onClick: onRecordClick)
                .padding(.bottom, VoraSpacing.floatingActionBottom)
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(screenBackground)
    }

    private var screenBackground: Color { colorScheme == .dark ? .voraCharcoal : .voraPaper }
    private var titleColor: Color { colorScheme == .dark ? .voraPaper : .voraLogoInk }
    private var metaColor: Color { colorScheme == .dark ? Color(hex: 0xB8C1CD) : .voraMuted }
    private var searchBackground: Color { colorScheme == .dark ? Color(hex: 0x1B2432) : .voraWhite }
    private var searchBorder: Color { colorScheme == .dark ? Color(hex: 0x24314F) : Color(hex: 0xE7EDF3) }
    private var cardBackground: Color { colorScheme == .dark ? Color(hex: 0x161F2D) : .voraWhite }
    private var cardBorder: Color { colorScheme == .dark ? Color(hex: 0x202C3D) : Color(hex: 0xEDF2F5) }
    private var emptyStateBackground: Color { colorScheme == .dark ? Color(hex: 0x161F2D) : Color(hex: 0xF7F9FB) }
    private var emptyStateBorder: Color { colorScheme == .dark ? Color(hex: 0x202C3D) : Color(hex: 0xF0F3F6) }
    private var emptyStateIconColor: Color { colorScheme == .dark ? Color(hex: 0xB8C1CD) : .voraLogoInk.opacity(0.78) }
    private var sourceChipBackground: Color { colorScheme == .dark ? Color(hex: 0x1D2736) : Color(hex: 0xEFF4F7) }
    private var sourceChipText: Color { colorScheme == .dark ? .voraPaper : .voraLogoInk }
    private var statusChipAccent: Color { colorScheme == .dark ? .voraTertiary : Color(hex: 0x3A7D58) }
}

#Preview("Notes Loading") {
    NotesScreen(
        mode: .loading,
        summaryText: "Loading local library",
        statusLabel: "Syncing",
        filters: [],
        memos: [],
        onRecordClick: {},
        onSelectSourceFilter: { _ in },
        onRenameMemo: { _ in },
        onDeleteMemo: { _ in }
    )
}

#Preview("Notes Loading Dark") {
    NotesScreen(
        mode: .loading,
        summaryText: "Loading local library",
        statusLabel: "Syncing",
        filters: [],
        memos: [],
        onRecordClick: {},
        onSelectSourceFilter: { _ in },
        onRenameMemo: { _ in },
        onDeleteMemo: { _ in }
    )
    .preferredColorScheme(.dark)
}

#Preview("Notes Empty") {
    NotesScreen(
        mode: .empty,
        summaryText: "0 memos",
        statusLabel: "Synced",
        filters: previewFilters(selected: "all"),
        memos: [],
        onRecordClick: {},
        onSelectSourceFilter: { _ in },
        onRenameMemo: { _ in },
        onDeleteMemo: { _ in }
    )
}

#Preview("Notes Empty Dark") {
    NotesScreen(
        mode: .empty,
        summaryText: "0 memos",
        statusLabel: "Synced",
        filters: previewFilters(selected: "all"),
        memos: [],
        onRecordClick: {},
        onSelectSourceFilter: { _ in },
        onRenameMemo: { _ in },
        onDeleteMemo: { _ in }
    )
    .preferredColorScheme(.dark)
}

#Preview("Notes Loaded") {
    NotesScreen(
        mode: .loaded,
        summaryText: "12 memos · 18 min",
        statusLabel: "Synced",
        filters: previewFilters(selected: "all"),
        memos: [
            MemoListItem(id: "memo-morning-idea", title: "Morning idea", time: "1:24", subtitle: "Today, 8:42", source: "Phone"),
            MemoListItem(id: "memo-pickup-notes", title: "Pickup notes", time: "0:48", subtitle: "Today, 12:10", source: "Watch")
        ],
        onRecordClick: {},
        onSelectSourceFilter: { _ in },
        onRenameMemo: { _ in },
        onDeleteMemo: { _ in }
    )
}

#Preview("Notes Loaded Dark") {
    NotesScreen(
        mode: .loaded,
        summaryText: "12 memos · 18 min",
        statusLabel: "Synced",
        filters: previewFilters(selected: "all"),
        memos: [
            MemoListItem(id: "memo-morning-idea", title: "Morning idea", time: "1:24", subtitle: "Today, 8:42", source: "Phone"),
            MemoListItem(id: "memo-pickup-notes", title: "Pickup notes", time: "0:48", subtitle: "Today, 12:10", source: "Watch")
        ],
        onRecordClick: {},
        onSelectSourceFilter: { _ in },
        onRenameMemo: { _ in },
        onDeleteMemo: { _ in }
    )
    .preferredColorScheme(.dark)
}

private func previewFilters(selected: String) -> [NotesSourceFilterItem] {
    [
        NotesSourceFilterItem(key: "all", label: "All", count: 12, selected: selected == "all"),
        NotesSourceFilterItem(key: "phone", label: "Phone", count: 7, selected: selected == "phone"),
        NotesSourceFilterItem(key: "smart", label: "Smart", count: 3, selected: selected == "smart"),
        NotesSourceFilterItem(key: "car", label: "Car", count: 2, selected: selected == "car")
    ]
}

File: composeApp/src/androidMain/kotlin/com/mohamedfaridelsherbini/vora/presentation/notes/NotesScreen.kt
package com.mohamedfaridelsherbini.vora.presentation.notes

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import com.mohamedfaridelsherbini.vora.presentation.theme.VoraSpacing

@Composable
internal fun NotesScreen(state: NotesListVisualState) {
    NotesScreen(
        state = state,
        onRecordClick = {},
        onSearchQueryChange = {},
        onSelectSourceFilter = {},
        onRenameRequest = {},
        onDeleteRequest = {},
        onRenameConfirm = {},
        onDeleteConfirm = {},
        onDismissRename = {},
        onDismissDelete = {},
    )
}

@Composable
internal fun NotesScreen(
    state: NotesListVisualState,
    onRecordClick: () -> Unit,
    onSearchQueryChange: (String) -> Unit,
    onSelectSourceFilter: (String) -> Unit,
    onRenameRequest: (String) -> Unit,
    onDeleteRequest: (String) -> Unit,
    onRenameConfirm: (String) -> Unit,
    onDeleteConfirm: () -> Unit,
    onDismissRename: () -> Unit,
    onDismissDelete: () -> Unit,
) {
    Box(
        modifier = Modifier
            .fillMaxSize()
            .background(state.background)
            .padding(horizontal = VoraSpacing.PageHorizontal),
    ) {
        NotesListContent(
            modifier = Modifier.fillMaxSize(),
            state = state,
            onSearchQueryChange = onSearchQueryChange,
            onSelectSourceFilter = onSelectSourceFilter,
            onRenameRequest = onRenameRequest,
            onDeleteRequest = onDeleteRequest,
        )

        RecordFab(
            modifier = Modifier
                .align(Alignment.BottomCenter)
                .padding(bottom = VoraSpacing.FloatingActionBottom),
            background = state.recordButtonBackground,
            textColor = state.recordButtonTextColor,
            onClick = onRecordClick,
        )
    }

    state.renameDialog?.let { dialog ->
        VoraRenameBottomSheet(
            memo = dialog.memo,
            onConfirm = onRenameConfirm,
            onDismiss = onDismissRename,
        )
    }

    state.deleteDialog?.let { dialog ->
        VoraDeleteDialog(
            memo = dialog.memo,
            onConfirm = onDeleteConfirm,
            onDismiss = onDismissDelete,
        )
    }
}

@Composable
private fun NotesListContent(
    modifier: Modifier = Modifier,
    state: NotesListVisualState,
    onSearchQueryChange: (String) -> Unit,
    onSelectSourceFilter: (String) -> Unit,
    onRenameRequest: (String) -> Unit,
    onDeleteRequest: (String) -> Unit,
) {
    LazyColumn(
        modifier = modifier,
        verticalArrangement = Arrangement.spacedBy(NotesScreenDefaults.SectionSpacing),
        contentPadding = PaddingValues(
            top = VoraSpacing.PageTop,
            bottom = NotesScreenDefaults.ListBottomContentPadding,
        ),
    ) {
        item {
            NotesHeader(
                titleColor = state.titleColor,
                subtitleColor = state.subtitleColor,
                summaryText = state.summaryText,
                statusLabel = state.statusLabel,
                statusBackground = state.statusChipBackground,
                statusTextColor = state.statusChipTextColor,
            )
        }
        item {
            NotesSearchBar(
                query = state.searchQuery,
                onQueryChange = onSearchQueryChange,
                background = state.searchBackground,
                borderColor = state.searchBorder,
                textColor = state.searchTextColor,
                iconColor = state.metaColor,
            )
        }
        when (state.mode) {
            NotesListMode.Loading -> loadingContent(state)
            NotesListMode.Empty -> emptyContent(state)
            NotesListMode.Loaded -> loadedContent(
                state = state,
                onSelectSourceFilter = onSelectSourceFilter,
                onRenameRequest = onRenameRequest,
                onDeleteRequest = onDeleteRequest,
            )
        }
    }
}

private fun androidx.compose.foundation.lazy.LazyListScope.loadingContent(
    state: NotesListVisualState,
) {
    item {
        NotesLoadingFilterRow(
            background = state.filterChipBackground,
            selectedBackground = state.titleColor,
        )
    }
    item {
        NotesLoadingStatusCard(
            titleColor = state.titleColor,
            subtitleColor = state.metaColor,
            background = state.cardBackground,
            borderColor = state.cardBorder,
            iconTint = state.metaColor,
            accentColor = state.statusChipBackground,
            skeletonColor = state.filterChipBackground,
        )
    }
    items(count = 4) {
        LoadingMemoCard(
            background = state.cardBackground,
            borderColor = state.cardBorder,
            lineColor = state.searchBorder,
            chipColor = state.filterChipBackground,
        )
    }
}

private fun androidx.compose.foundation.lazy.LazyListScope.emptyContent(
    state: NotesListVisualState,
) {
    item {
        NotesEmptyStateCard(
            modifier = Modifier.padding(top = 96.dp),
            titleColor = state.titleColor,
            subtitleColor = state.metaColor,
            background = state.emptyStateBackground,
            borderColor = state.emptyStateBorder,
            iconColor = state.emptyStateIconColor,
        )
    }
}

private fun androidx.compose.foundation.lazy.LazyListScope.loadedContent(
    state: NotesListVisualState,
    onSelectSourceFilter: (String) -> Unit,
    onRenameRequest: (String) -> Unit,
    onDeleteRequest: (String) -> Unit,
) {
    item {
        NotesFilterRow(
            filters = state.filters,
            background = state.filterChipBackground,
            textColor = state.filterChipTextColor,
            onSelect = onSelectSourceFilter,
        )
    }
    item {
        NotesRecentLabel(
            textColor = state.metaColor,
        )
    }
    items(
        items = state.memos,
        key = { memo -> memo.id },
    ) { memo ->
        SwipeRevealItem(
            onRename = { onRenameRequest(memo.id) },
            onDelete = { onDeleteRequest(memo.id) },
        ) {
            MemoCard(
                memo = memo,
                titleColor = state.titleColor,
                metaColor = state.metaColor,
                cardBackground = state.cardBackground,
                cardBorder = state.cardBorder,
                sourceBackground = state.filterChipBackground,
                sourceTextColor = state.filterChipTextColor,
            )
        }
    }
}

private object NotesScreenDefaults {
    val SectionSpacing = 14.dp
    val ListBottomContentPadding = 104.dp
}

@Preview(name = "Phone Notes", showBackground = true, backgroundColor = 0xFFF9FAFB)
@Composable
private fun NotesScreenPreview() {
    NotesScreen(
        state = previewLoadedNotesUiState().toVisualState(dark = false),
    )
}

@Preview(name = "Phone Notes Dark", showBackground = true, backgroundColor = 0xFF111827)
@Composable
private fun NotesScreenDarkPreview() {
    NotesScreen(
        state = previewLoadedNotesUiState().toVisualState(dark = true),
    )
}

@Preview(name = "Phone Notes Empty", showBackground = true, backgroundColor = 0xFFF9FAFB)
@Composable
private fun NotesScreenEmptyPreview() {
    NotesScreen(
        state = previewEmptyNotesUiState().toVisualState(dark = false),
    )
}

@Preview(name = "Phone Notes Empty Dark", showBackground = true, backgroundColor = 0xFF111827)
@Composable
private fun NotesScreenEmptyDarkPreview() {
    NotesScreen(
        state = previewEmptyNotesUiState().toVisualState(dark = true),
    )
}

@Preview(name = "Phone Notes Loading", showBackground = true, backgroundColor = 0xFFF9FAFB)
@Composable
private fun NotesScreenLoadingPreview() {
    NotesScreen(
        state = previewLoadingNotesUiState().toVisualState(dark = false),
    )
}

@Preview(name = "Phone Notes Loading Dark", showBackground = true, backgroundColor = 0xFF111827)
@Composable
private fun NotesScreenLoadingDarkPreview() {
    NotesScreen(
        state = previewLoadingNotesUiState().toVisualState(dark = true),
    )
}
